File: src/main/java/be/hubau/sane/parser/ParserException.java
package be.hubau.sane.parser;

import java.util.Stack;

/**
 * <p>Signals that a parse failure has occurred.</p>
 *
 * <p>Producer : com.parse2.aparse.Parser 2.5<br/>
 * Produced : Mon Apr 15 15:26:09 CEST 2019</p>
 */

public class ParserException extends Exception {
    static final private String newline = System.getProperty("line.separator", "\n");
    private String reason;
    private String text60;
    private int index60;
    private Stack<String> ruleStack;

    /**
     * Creates a parser exception from the specified parse failure information.
     *
     * @param reason    A description of the parse failure.
     * @param text      The string of characters being parsed.
     * @param index     The index to the character at which the parse failure occurred.
     * @param ruleStack The ABNF rule stack at the point the parse failure occurred.
     */

    public ParserException(String reason, String text, int index, Stack<String> ruleStack) {
        this.reason = reason;
        this.ruleStack = ruleStack;

        int start = (index < 30) ? 0 : index - 30;
        int end = (text.length() < index + 30) ? text.length() : index + 30;
        this.text60 = text.substring(start, end).replaceAll("[\\x00-\\x1F]", " ");
        this.index60 = (index < 30) ? index : 30;
    }

    public String getSubstring() {
        return this.text60;
    }

    public String getMessage() {
        String marker = "                              ";

        StringBuilder buffer = new StringBuilder();
        buffer.append(this.reason).append(newline);
        buffer.append(this.text60).append(newline);
        buffer.append(marker, 0, this.index60).append("^").append(newline);

        if (!this.ruleStack.empty()) {
            buffer.append("rule stack:");

            for (String rule : this.ruleStack)
                buffer.append(newline).append("  ").append(rule);
        }

        ParserException secondaryError = (ParserException) getCause();
        if (secondaryError != null) {
            buffer.append("possible cause: ").append(secondaryError.reason).append(newline);
            buffer.append(secondaryError.text60).append(newline);
            buffer.append(marker.substring(0, secondaryError.index60)).append("^").append(newline);

            if (!secondaryError.ruleStack.empty()) {
                buffer.append("rule stack:");

                for (String rule : secondaryError.ruleStack)
                    buffer.append(newline).append("  ").append(rule);
            }
        }

        return buffer.toString();
    }
}

File: src/main/java/be/hubau/sane/parser/ParserException.kt
package be.hubau.sane.parser

import java.util.*

/**
 * Signals that a parse failure has occurred.
 */

class ParserException

/**
 * Creates a parser exception from the specified parse failure information.
 *
 * @param reason A description of the parse failure.
 * @param text The string of characters being parsed.
 * @param index The index to the character at which the parse failure occurred.
 * @param ruleStack The ABNF rule stack at the point the parse failure occurred.
 */
(private val reason: String, text: String, index: Int, private val ruleStack: Stack<String>) : BaseException() {
    /**
     * Returns a substring of the parsed string that encompasses the point
     * at which the parse failure occurred. The substring will be up to 60
     * characters in length unless the point of failure occurred within
     * 30 characters of the start or end of the parsed string.
     * [.getSubstringIndex] returns an index to the character within
     * this substring at which the parse failure occurred. This substring
     * may contain non-printable characters.
     *
     * @return The substring that encompasses the point of failure.
     */
    val substring: String

    /**
     * Returns an index to the character within the substring returned by
     * [.getSubstring] at which the parse failure occurred.
     *
     * @return The index to the character within the substring returned
     * [.getSubstring] at which the parse failure occurred.
     */
    private val substringIndex: Int

    init {
        val start = if (index < 30) 0 else index - 30
        val end = if (text.length < index + 30) text.length else index + 30
        substring = text.substring(start, end).replace("[\\x00-\\x1F]".toRegex(), " ")
        substringIndex = if (index < 30) index else 30
    }

    override fun errorMessage(): String {
        val marker = "                              "

        val buffer = StringBuffer()
        buffer.append(reason + newline)
        buffer.append(substring + newline)
        buffer.append(marker.substring(0, substringIndex) + "^" + newline)

        if (!ruleStack.empty()) {
            buffer.append("rule stack:")

            for (rule in ruleStack)
                buffer.append("$newline  $rule")
        }

        if (cause != null) {
            val secondaryError = cause as ParserException
            buffer.append("possible cause: " + secondaryError.reason + newline)
            buffer.append(secondaryError.substring + newline)
            buffer.append(marker.substring(0, secondaryError.substringIndex) + "^" + newline)

            if (!secondaryError.ruleStack.empty()) {
                buffer.append("rule stack:")

                for (rule in secondaryError.ruleStack)
                    buffer.append("$newline  $rule")
            }
        }

        return buffer.toString()
    }

    companion object {
        private val newline = System.getProperty("line.separator", "\n")
    }
}